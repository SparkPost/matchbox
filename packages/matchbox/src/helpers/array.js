/**
 * Group a list of items into sub groups based on the value of each item's groupingKey
 *
 * @param {Array} list List of items to group
 * @param {String} groupingKey Key to use for grouping each item by this value
 *
 * @example
 * const list = [{ name: 'a', section: 1 }, { name: 'b', section: 2 }, { name: 'c', section: 1 }]
 *
 * groupBy(list, 'section')
 *
 * // returns
 * {
 *   1: [ { name: 'a', section: 1 }, { name: 'c', section: 1 } ],
 *   2: [ { name: 'b', section: 2 } ]
 * }
 */
export const groupBy = (list, groupingKey) => list.reduce((acc, item) => {
  const index = item[groupingKey];
  return { ...acc, [index]: [ ...(acc[index] || []), item ]};
}, {});

export const groupByValues = (list, groupingKey) => Object.values(groupBy(list, groupingKey));

/**
 * Filters by an array of actions by the visible key
 * @param {Array} actions - Array of actions
 *
 * @example
 * filterVisible([{ name: 'a', visible: true }, { name: 'b', visible: false }, { name: 'c' }])
 *
 * // returns
 * [{ name: 'a' }, { name: 'c' }]
 */
export const filterByVisible = (actions) => actions.reduce((acc, { visible = true, ...action }) => {
  if (visible) {
    acc.push(action);
  }
  return acc;
}, []);
