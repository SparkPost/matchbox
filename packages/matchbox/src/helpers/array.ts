/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * Group a list of items into sub groups based on the value of each item's groupingKey
 *
 * @param {Array} list List of items to group
 * @param {String} groupingKey Key to use for grouping each item by this value
 *
 * @example
 * const list = [{ name: 'a', section: 1 }, { name: 'b', section: 2 }, { name: 'c', section: 1 }]
 *
 * groupBy(list, 'section')
 *
 * // returns
 * {
 *   1: [ { name: 'a', section: 1 }, { name: 'c', section: 1 } ],
 *   2: [ { name: 'b', section: 2 } ]
 * }
 */
export const groupBy = (list: any[], groupingKey: string): any[] =>
  list.reduce((acc, item) => {
    const index = item[groupingKey];
    return { ...acc, [index]: [...(acc[index] || []), item] };
  }, {});

export const groupByValues = (list: any[], groupingKey: string): any[] =>
  Object.values(groupBy(list, groupingKey));

/**
 * Filters by an array of actions by the visible key
 * @param {Array} actions - Array of actions
 *
 * @example
 * filterVisible([{ name: 'a', visible: true }, { name: 'b', visible: false }, { name: 'c' }])
 *
 * // returns
 * [{ name: 'a' }, { name: 'c' }]
 */
export const filterByVisible = (actions: any[]): any[] => {
  if (!actions) {
    return [];
  }

  return actions.reduce((acc, { visible = true, ...action }) => {
    if (visible) {
      acc.push(action);
    }
    return acc;
  }, []);
};
