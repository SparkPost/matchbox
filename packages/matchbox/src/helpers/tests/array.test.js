import { groupBy, groupByValues, filterByVisible } from '../array';

describe('Helper: Array helpers', () => {

  let list;

  beforeEach(() => {
    list = [
      {
        name: 'a',
        section: 1
      },
      {
        name: 'b',
        section: 4
      },
      {
        name: 'c',
        section: 1
      },
      {
        name: 'd',
        section: 'mixed whoa'
      }
    ];
  });

  describe('groupBy', () => {

    it('should group a list by a given key', () => {
      expect(groupBy(list, 'section')).toMatchSnapshot();
    });

    it('should group items without a key into an undefined group', () => {
      list.push({ name: 'no section 1' });
      list.push({ name: 'no section 2' });
      const grouped = groupBy(list, 'section');
      expect(grouped).toMatchSnapshot();
      expect(grouped.undefined).toHaveLength(2);
    });

    it('should group items into a single group when the group by key is not present in any item', () => {
      const grouped = groupBy(list, 'zerp');
      expect(Object.keys(grouped)).toHaveLength(1);
      expect(grouped).toMatchSnapshot();
    });

  });

  describe('groupByValues', () => {

    it('should return a grouped array', () => {
      expect(groupByValues(list, 'section')).toMatchSnapshot();
    });

  });

  describe('filterByVisible', () => {
    it('should return only visible actions', () => {
      expect(filterByVisible([
        { content: '1' },
        { content: '2', visible: true },
        { content: '3', visible: false }
      ])).toMatchSnapshot();
    });
  });

});
