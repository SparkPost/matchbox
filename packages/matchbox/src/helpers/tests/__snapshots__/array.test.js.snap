// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper: Array helpers filterByVisible should return only visible actions 1`] = `
Array [
  Object {
    "content": "1",
  },
  Object {
    "content": "2",
  },
]
`;

exports[`Helper: Array helpers groupBy should group a list by a given key 1`] = `
Object {
  "1": Array [
    Object {
      "name": "a",
      "section": 1,
    },
    Object {
      "name": "c",
      "section": 1,
    },
  ],
  "4": Array [
    Object {
      "name": "b",
      "section": 4,
    },
  ],
  "mixed whoa": Array [
    Object {
      "name": "d",
      "section": "mixed whoa",
    },
  ],
}
`;

exports[`Helper: Array helpers groupBy should group items into a single group when the group by key is not present in any item 1`] = `
Object {
  "undefined": Array [
    Object {
      "name": "a",
      "section": 1,
    },
    Object {
      "name": "b",
      "section": 4,
    },
    Object {
      "name": "c",
      "section": 1,
    },
    Object {
      "name": "d",
      "section": "mixed whoa",
    },
  ],
}
`;

exports[`Helper: Array helpers groupBy should group items without a key into an undefined group 1`] = `
Object {
  "1": Array [
    Object {
      "name": "a",
      "section": 1,
    },
    Object {
      "name": "c",
      "section": 1,
    },
  ],
  "4": Array [
    Object {
      "name": "b",
      "section": 4,
    },
  ],
  "mixed whoa": Array [
    Object {
      "name": "d",
      "section": "mixed whoa",
    },
  ],
  "undefined": Array [
    Object {
      "name": "no section 1",
    },
    Object {
      "name": "no section 2",
    },
  ],
}
`;

exports[`Helper: Array helpers groupByValues should return a grouped array 1`] = `
Array [
  Array [
    Object {
      "name": "a",
      "section": 1,
    },
    Object {
      "name": "c",
      "section": 1,
    },
  ],
  Array [
    Object {
      "name": "b",
      "section": 4,
    },
  ],
  Array [
    Object {
      "name": "d",
      "section": "mixed whoa",
    },
  ],
]
`;
